processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,385 @@
+import io.druid.java.util.common.IAE;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.java.util.common.parsers.CloseableIterator;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.segment.ColumnSelectorFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.NoSuchElementException;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A streaming grouper which can aggregate sorted inputs.  This grouper can aggregate while its iterator is being
+ * consumed.  Also, the aggregation thread and iterating thread can be different.
+ *
+ * This grouper is backed by a circular array off-heap buffer.  Reading iterator is able to read from an array slot only
+ * if the write for that slot is finished.
+ */
+public class StreamingMergeSortedGrouper<KeyType> implements Grouper<KeyType>
+{
+  private static final Logger LOG = new Logger(StreamingMergeSortedGrouper.class);
+  // Timeout for waiting for a slot to be available for read/write. This is required to prevent for the processing
+  // thread from being blocked if its iterator is not consumed due to some failure.
+  private static final long DEFAULT_TIMEOUT_MS = 5000L;
+
+  private final Supplier<ByteBuffer> bufferSupplier;
+  private final KeySerde<KeyType> keySerde;
+  private final BufferAggregator[] aggregators;
+  private final int[] aggregatorOffsets;
+  private final int keySize;
+  private final int recordSize; // size of (key + all aggregates)
+  private final long timeoutMs;
+
+  // Below variables are initialized when init() is called
+  private ByteBuffer buffer;
+  private int maxSlotNum;
+  private boolean initialized;
+
+  // Below variables can be read/written by differernt threads
+
+  /**
+   * Indicate that this grouper consumed the last input or not.
+   */
+  private volatile boolean finished;
+
+  /**
+   * Currently writing position. This is always moved ahead of nextReadIndex.
+   */
+  private volatile int curWriteIndex;
+
+  /**
+   * Next read position. This can be moved to a position only when write for the position is finished.
+   */
+  private volatile int nextReadIndex;
+
+  /**
+   * Returns the minimum buffer capacity required to use this grouper.  This grouper keeps track read/write indexes
+   * and they cannot point the same position at the same time.  Since the read/write indexes circularly, the required
+   * minimum buffer capacity is 3 * record size.
+   *
+   * @return required minimum buffer capacity
+   */
+  public static <KeyType> int requiredBufferCapacity(
+      KeySerde<KeyType> keySerde,
+      AggregatorFactory[] aggregatorFactories
+  )
+  {
+    int recordSize = keySerde.keySize();
+    for (AggregatorFactory aggregatorFactory : aggregatorFactories) {
+      recordSize += aggregatorFactory.getMaxIntermediateSize();
+    }
+    return recordSize * 3;
```

Comment:
What is 3?

Reply:
This is explained in java doc (https://github.com/druid-io/druid/pull/4704/files/b748da59631fb3c5a881a6eefb61826ed39cdf24#diff-8bb57c81d0c016c3203d4b3af9b741a4R83).